main.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"flag"
+	"log"
+	"net/http"
+	"os/user"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/tkanos/gonfig"
+)
+
+const (
+	DOMAIN_KEY        = "domain"
+	CONFIG            = "catchall.json"
+	StatusUnknown     = "unknown"
+	StatusCatchAll    = "catch-all"
+	StatusNotCatchAll = "not catch-all"
+	CatchAllTrigger   = 1000
+	HasBounced        = -1
+	UnrecordedBounce  = -2
+	RetryInsertTime   = 60 // seconds
+)
+
+func serveHttp(
+	wg *sync.WaitGroup, queue Queue,
+	backendRead BackendReader, port int) {
+	defer wg.Done()
+	router := mux.NewRouter()
+	// Write mode (PUTs)
+	router.Handle(
+		"/events/{"+DOMAIN_KEY+"}/delivered",
+		HandleRequest(false, queue)).Methods("PUT")
+	router.Handle(
+		"/events/{"+DOMAIN_KEY+"}/bounced",
+		HandleRequest(true, queue)).Methods("PUT")
+	router.Handle(
+		"/stats/{"+DOMAIN_KEY+"}",
+		HandleStatsQuery(queue)).Methods("GET")
+	router.Handle(
+		"/domains/{"+DOMAIN_KEY+"}",
+		HandleQuery(backendRead)).Methods("GET")
+	router.PathPrefix("/").HandlerFunc(CatchAllHandler)
+
+	// NOTE(roaet): Possible performance increase with valyala/fasthttp
+	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(port), router))
+}
+
+func doWork(
+	wg *sync.WaitGroup, thisHost string, queue Queue,
+	backend Backend) {
+	defer wg.Done()
+	for {
+		entry := HostEntry{Host: thisHost}
+		backend.ping(entry)
+		err := queue.persist()
+		if err != nil {
+			log.Println("Error persisting queue")
+		}
+		_, err = backend.update()
+		if err != nil {
+
+		}
+		time.Sleep(10 * time.Second)
+	}
+}
+
+type Configuration struct {
+	Port               int
+	Worker, API, Local bool
+	APIMode, Host      string
+	CassandraHosts     []string
+	RabbitURI          string
+}
+
+func initializeConfig(filepath string, config *Configuration) error {
+	usr, err := user.Current()
+	if err != nil {
+		log.Fatalf("Could not load user information: %s\n", err)
+	}
+	homedir := usr.HomeDir
+	err = gonfig.GetConf(homedir+"/"+CONFIG, config)
+	if err != nil {
+		log.Fatalf("Could not load config: %s: %s\n", CONFIG, err)
+	}
+	return nil
+}
+
+func main() {
+	var port int
+	var host string
+	var queue Queue
+	var backend Backend
+	var wg sync.WaitGroup
+
+	configuration := Configuration{}
+	initializeConfig(CONFIG, &configuration)
+
+	flag.IntVar(&port, "port", configuration.Port, "port to listen on")
+	flag.StringVar(&host, "host", configuration.Host, "host ip to listen on")
+	flag.Parse()
+	log.Printf("Host: %s\n", host)
+	log.Printf("Port: %d\n", port)
+	host_str := host + ":" + strconv.Itoa(port)
+	hosts := []string{host_str}
+	cass := NewCassandraBackend(configuration.CassandraHosts)
+	defer cass.EndSession()
+	httpClient := &DefaultHTTP{}
+	intercom := NewHTTPIntercom(httpClient)
+	backend, _ = NewDistributedBackend(hosts, cass, intercom)
+	queue = NewCachedQueue(cass)
+
+	wg.Add(1)
+	go doWork(&wg, host_str, queue, backend)
+	wg.Add(1)
+	go serveHttp(&wg, queue, backend, port)
```

Comment:
How could we further scale this on a single node?

Reply:
Many more goroutines and probably a task list (buffered channel or something similar). I haven't had the chance to run this through pprof or trace but I imagine that there's a lot of waiting here. There is a more efficient http implementation as well that is 10x faster than the built in one.

I also think this could be a state machine (or just event driven) to make it more flexible. Such a thing would allow better retries on the 'persist to backend' instead of the current delay model.

Reply:
I ran pprof on it and found that json.Marshal is taking up pretty decent portion of the time. A quick gain would come from just constructing a json string instead of using Marshal.